### PR TITLE
Definition link API

### LINKED ISSUE #10037: Extend language API to allow sourceRange in Go to Definition
- VSCode Version: Code 1.3.1 (e6b4afa53e9c0f54edef1673de9001e9f0f547ae, 2016-07-12T13:35:06.227Z)
- OS Version: Windows_NT ia32 6.1.7601

Steps to Reproduce:
1. Create a typescript project for eg angular2
2. create a component with name having dot character
3. import the module in to another component. 

ctrl + hover - highlighting upto dot character
in the below issue the ctl + hover should highlight **gateway.destination.component** but it is highlighting only the _component_ alone. 
vscode-ctrl-hover

### PR DESCRIPTION
**Pull request by mjbvz**
Add a new `DefinitionLink` type. This type allows definition providers to return additional metadata about a definition, such as the defining span.

Hook up this new provider for typescript

This PR replaces #48001

Fixes #10037

### CODE REVIEW

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -538,4 +538,19 @@ declare module 'vscode' {
...
 
 	//#endregion
+
+	//#region Matt: Deinition range
+
+	export interface DefinitionLink {
+		origin?: Range;
+		uri: Uri;
+		range: Range;
+		selectionRange?: Range;
```
Comment (mjbvz) [inline src/vs/vscode.proposed.d.ts]: using these field names make `DefinitionLink` compatible with `Location`

Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: sneaky... we could also make `origin` mandatory and use the `position` of the request as `origin`

Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: or did you plan that as non-breaking API evolution?

**Comment by jrieken** [comment]
@mjbvz Please fix merge conflicts

File: src/vs/vscode.proposed.d.ts (hunk 1 above)
Comment (mjbvz) [reply src/vs/vscode.proposed.d.ts]: Main reason for making it optional is that I think it makes sense for implementers:

- When `origin` is provided: always use this value.
- When `origin` is undefined: use the word range at the request position as the `origin`

If `origin` is not optional, I believe many extensions would have to compute the word range themselves

File: src/vs/workbench/api/node/extHostLanguageFeatures.ts [hunk 2]
```diff
@@ -150,18 +150,33 @@ class DefinitionAdapter {
...
+				return (value as (vscode.DefinitionLink | vscode.Location)[]).map(x => DefinitionAdapter.convertDefinitionLink(position, x));
 			} else if (value) {
-				return typeConvert.location.from(value);
+				return [DefinitionAdapter.convertDefinitionLink(position, value)];
 			}
 			return undefined;
 		});
 	}
+
+	private static convertDefinitionLink(position: IPosition, value: vscode.Location | vscode.DefinitionLink): modes.DefinitionLink {
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostLanguageFeatures.ts]: `position` is unused. move to extHostTypeConverter?

File: src/vs/vscode.proposed.d.ts [hunk 3]
```diff
@@ -681,4 +681,19 @@ declare module 'vscode' {
...
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts]: One a second thought I wonder if defaulting `DefinitionLink.range` to `Location.range` is the right thing... It feels like its the `selectionRange` cos that how it is currently used. It's also OK to leave this out of this PR but something we should likely do when we optimise how the preview hover is computed.

File: src/vs/editor/contrib/goToDefinition/goToDefinitionMouse.ts [hunk 4]
```diff
@@ -157,8 +157,16 @@ class GotoDefinitionWithMouseEditorContribution implements editorCommon.IEditorC
 					const previewRange = new Range(startLineNumber, 1, endLineNumber + 1, 1);
 					const value = textEditorModel.getValueInRange(previewRange).replace(new RegExp(`^\\s{${minIndent - 1}}`, 'gm'), '').trim();
 
+					let wordRange: Range;
+					if (result.origin) {
+						const range = result.origin;
+						wordRange = new Range(range.startLineNumber, range.startColumn, range.endLineNumber, range.endColumn);
```
Comment (jrieken) [inline src/vs/editor/contrib/goToDefinition/goToDefinitionMouse.ts]: fyi - `Range.lift` does the job

**Review by jrieken — APPROVED** [review]
nits - feel free to merge

**Comment by mjbvz** [comment]
@jrieken Thanks. I've made the suggested changes and will merge this into proposed

I may be confusing `range` and `selectionRange` though. For a class definition, should `range` be the full range of a class body, while selection range would be the range of the class name? Or is it the reverse?

**Comment by jrieken** [comment]
Yeah, `range` should be the full range but I think todays implementations return something that closer to the `selectionRange`

**Pull request merged**